File: src/ccx/graphs.py
"""Load JSON-LD graph files into an rdflib Dataset, with no network access."""

from __future__ import annotations

import json
import socket
from contextlib import contextmanager

import rdflib

from .errors import CCXSecurityError, CCXValidationError


@contextmanager
def no_network():
    """Block all outbound socket connections for the duration of the block.

    Patches ``socket.socket.connect`` process-wide; it is NOT thread-safe or
    reentrant. Reading a CCX package is synchronous, so a single guard wraps the
    whole parse. The reentrancy guard raises if the patch is already active.
    """
    original = socket.socket.connect
    if getattr(original, "_ccx_blocked", False):
        raise CCXSecurityError("no_network() is not reentrant")

    def blocked(self, *args, **kwargs):
        raise CCXSecurityError("network access blocked while reading package")

    blocked._ccx_blocked = True
    socket.socket.connect = blocked
    try:
        yield
    finally:
        socket.socket.connect = original


def graph_iri(namespace: str, name: str) -> str:
    return f"https://w3id.org/ccx/graph/{namespace}/{name}"


def reject_remote_context(doc: object) -> None:
    """Raise CCXSecurityError if a JSON-LD doc references a remote context anywhere.

    Walks the whole structure: a remote ``@context`` (string or list element) or a
    remote JSON-LD 1.1 ``@import`` is rejected whether it sits at the top level,
    inside ``@graph``, or in a node-scoped context nested anywhere.
    """

    def is_remote(value: object) -> bool:
        return isinstance(value, str) and value.startswith(("http://", "https://"))

    if isinstance(doc, dict):
        ctx = doc.get("@context")
        candidates = ctx if isinstance(ctx, list) else [ctx]
        for candidate in candidates:
            if is_remote(candidate):
                raise CCXSecurityError(
                    f"remote @context not allowed: {candidate!r} (bundle it in the package)"
                )
        imported = doc.get("@import")
        if is_remote(imported):
            raise CCXSecurityError(
                f"remote @import not allowed: {imported!r} (bundle it in the package)"
            )
        for value in doc.values():
            reject_remote_context(value)
    elif isinstance(doc, list):
        for item in doc:
            reject_remote_context(item)


def is_default(entry) -> bool:
    return entry.role == "default" or entry.namespace == "ccx"


def load_dataset(container, manifest) -> rdflib.Dataset:
    """Assemble the package's graphs into an rdflib Dataset.

    The `knowledge` graph (role 'default' or namespace 'ccx') becomes the RDF
    default graph; every other declared graph becomes a named graph.
    """
    dataset = rdflib.Dataset()
    with no_network():
        for entry in manifest.graphs:
            text = container.read(entry.path).decode("utf-8")
            reject_remote_context(json.loads(text))
            target = (
                dataset.default_graph
                if is_default(entry)
                else dataset.graph(rdflib.URIRef(graph_iri(entry.namespace, entry.name)))
            )
            try:
                target.parse(data=text, format="json-ld")
            except CCXSecurityError:
                raise
            except Exception as exc:  # noqa: BLE001 - surface a uniform error
                raise CCXValidationError(
                    f"failed to parse graph {entry.path!r}: {exc}"
                ) from exc
    return dataset

File: js/src/graphs.ts
/**
 * JSON-LD graph helpers for the CCX reader.
 *
 * Mirrors the network-safety and default-graph logic in `src/ccx/graphs.py`:
 * - `rejectRemoteContext` -> `reject_remote_context`
 * - `isDefault`           -> `is_default`
 *
 * The reader MUST NOT fetch remote JSON-LD contexts at read time, so any remote
 * `@context` or JSON-LD 1.1 `@import` reference is rejected up front.
 */

import { CcxSecurityError } from "./errors";

/** True when `value` is a string that looks like a remote (http/https) URL. */
function isRemote(value: unknown): value is string {
  return typeof value === "string" && /^https?:\/\//i.test(value);
}

/** True when `value` is a plain object (and not an array or null). */
function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/**
 * Throw `CcxSecurityError` if a JSON-LD doc references a remote context anywhere.
 *
 * Walks the whole structure: a remote `@context` (string or list element) or a
 * remote JSON-LD 1.1 `@import` is rejected whether it sits at the top level,
 * inside `@graph`, or in a node-scoped context nested anywhere.
 */
export function rejectRemoteContext(node: unknown): void {
  if (isRecord(node)) {
    const ctx = node["@context"];
    const candidates = Array.isArray(ctx) ? ctx : [ctx];
    for (const candidate of candidates) {
      if (isRemote(candidate)) {
        throw new CcxSecurityError(
          `remote @context not allowed: ${candidate} (bundle it in the package)`,
        );
      }
    }
    const imported = node["@import"];
    if (isRemote(imported)) {
      throw new CcxSecurityError(
        `remote @import not allowed: ${imported} (bundle it in the package)`,
      );
    }
    for (const value of Object.values(node)) {
      rejectRemoteContext(value);
    }
  } else if (Array.isArray(node)) {
    for (const item of node) {
      rejectRemoteContext(item);
    }
  }
}

/**
 * True when a graph entry maps to the RDF default graph: it is explicitly the
 * `default` role, or it lives in the reserved `ccx` namespace.
 */
export function isDefault(e: { role?: string | null; namespace: string }): boolean {
  return e.role === "default" || e.namespace === "ccx";
}
